promql/prettier/prettier.go

```diff
@@ -0,0 +1,480 @@
+					if items[j].Typ == parser.RIGHT_PAREN {
+						closingParenIndex = j
+						break
+					}
+				}
+				if closingParenIndex == -1 {
+					panic("invalid paren index: closing paren not found")
+				}
+				// Re-order lexical items. TODO: consider using slicing of lists.
+				for itemp := 0; itemp < len(items); itemp++ {
+					if itemp <= aggregatorIndex || itemp > closingParenIndex {
+						tempItems = append(tempItems, items[itemp])
+					} else if !finishedKeywordWriting {
+						// Immediately after writing the aggregator, write the keyword and
+						// the following expression till the closing paren. This is done
+						// to be in-order with the expected result.
+						for j := keywordIndex; j <= closingParenIndex; j++ {
+							tempItems = append(tempItems, items[j])
+						}
+						// itemp has not been updated yet and its holding the left paren.
+						tempItems = append(tempItems, items[itemp])
+						finishedKeywordWriting = true
+					} else if !(itemp >= keywordIndex && itemp <= closingParenIndex) {
+						tempItems = append(tempItems, items[itemp])
+					}
+				}
+				items = tempItems
+			}
+
+		case parser.IDENTIFIER:
+			//	TODO: currently its assumed that STRING are always label values.
+			itr := advanceComments(items, i)
+			if i < 1 || item.Val != "__name__" || items[itr+2].Typ != parser.STRING {
+				continue
+			}
+			var (
+				leftBraceIndex  = -1
+				rightBraceIndex = -1
+				labelValItem    = items[itr+2]
+				metricName      = labelValItem.Val[1 : len(labelValItem.Val)-1] // Trim inverted commas.
+				tmp             []parser.Item
+				skipBraces      bool // For handling metric_name{} -> metric_name
+			)
+			// We aim to convert __name__ to an ideal metric_name if the value of that labels is atomic.
+			// Since a non-atomic metric_name will contain alphabets other than a-z and A-Z including _,
+			// anything that violates this ceases the formatting of that particular label item.
+			// If this is not done then the output from the prettier might be an un-parsable expression.
+			if !regexp.MustCompile("^[a-z_A-Z]+$").MatchString(metricName) {
+				continue
+			}
+			for backScanIndex := itr; backScanIndex >= 0; backScanIndex-- {
+				if items[backScanIndex].Typ == parser.LEFT_BRACE {
+					leftBraceIndex = backScanIndex
+					break
+				}
+			}
+			for forwardScanIndex := itr; forwardScanIndex < len(items); forwardScanIndex++ {
+				if items[forwardScanIndex].Typ == parser.RIGHT_BRACE {
+					rightBraceIndex = forwardScanIndex
+					break
+				}
+			}
+			// TODO: assuming comments are not present at this place.
+			itr = advanceComments(items, itr)
+			if items[itr+3].Typ == parser.COMMA {
+				skipBraces = rightBraceIndex-5 == advanceComments(items, leftBraceIndex)
+			} else {
+				skipBraces = rightBraceIndex-4 == advanceComments(items, leftBraceIndex)
+			}
+			identifierItem := parser.Item{Typ: parser.IDENTIFIER, Val: metricName, Pos: 0}
+			for j := 0; j < len(items); j++ {
```

Comment:
This whole thing needs more explanation how it works.

It would be even better if the code was drastically simplified.

Reply:
Sure. Will optimize and report back to you.

Reply:
I have added explanatory comments. Please have a look and comment on its satisfactoriness.

Reply:
While you described the algorithm, which could be inferred from the code, it is still not clear, what that whole functions is trying to achieve and why you chose to implement it the way you did.

Reply:
The explanatory comments for this case have been improved. Please have a look over it and comment on their satisfactoriness.